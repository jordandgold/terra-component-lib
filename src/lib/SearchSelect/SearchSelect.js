import React, { Component } from "react";
import "./SearchSelect.scss";

class SearchSelect extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selections: props.selections,
      deployed: false,
      searchField: "",
      selection: ""
    };
  }

  filterSelections = e => {
    this.setState({
      search: e.target.value
    });

    return this.state.selections.filter(selection => {
      return (
        selection.toLowerCase().substring(0, e.target.value.length - 1) ===
        e.target.value.toLowerCase()
      );
    });
  };

  renderOptions = () => {
    return this.state.selections.map((selection, index) => {
      <li key={`${selection}-${index}`}>{selection}</li>;
    });
  };

  toggleSelect = () => {
    this.setState({
      deployed: !this.state.deployed
    });
  };
}
