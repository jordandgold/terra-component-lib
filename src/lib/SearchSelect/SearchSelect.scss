@import "../global-styles/terra.scss";

.ter-search-select__select-options.drop-undefined {
  display: none;
}

.ter-search-select__select-options.drop-undefined.is-open {
  display: block;
}

.ter-search-select {
  display: block;
  position: relative;
  outline: none;
  &:hover {
    cursor: pointer;
  }
  &__selected {
    @include input-base();
    outline: none;
    position: relative;
  }
  &__caret {
    position: absolute;
    right: spacing("xs");
    top: 50%;
    transform: translateY(-50%);
  }
  &__options-list {
    border: $dropdown-border-size solid $dropdown-border-color;
    position: absolute;
    top: calc(100% + #{$dropdown-border-size});
    padding: 0;
    left: 0;
    list-style: none;
    background: $dropdown-bg;
    width: 100%;
    margin: 0;
    box-sizing: border-box;
    display: none;
    max-height: 20em;
    overflow: scroll;
    .is-open & {
      display: block;
    }
  }
  &__options-list-item {
    @include type-setting();
    font-size: $dropdown-item-font-size;
    color: $dropdown-link-color;
    padding: spacing("xxs") spacing("xs");
    display: block;
    text-decoration: none;
    &:hover {
      color: darken($dropdown-link-color, 10%);
      background: $dropdown-link-hover-bg;
      text-decoration: none;
    }
  }
  &__options-list-item--inactive {
    &:hover {
      cursor: default;
      background: transparent;
    }
  }
  &__search-input {
    @include input-base();
    background: white;
    border-top: 0px;
    border-left: 0px;
    border-right: 0px;
    border-color: color("gray-4");
  }
}
