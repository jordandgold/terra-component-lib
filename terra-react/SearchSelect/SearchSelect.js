import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import React, { Component } from "react";
import "./SearchSelect.scss";
import Icon from "../Icon/Icon";

var SearchSelect =
/*#__PURE__*/
function (_Component) {
  _inherits(SearchSelect, _Component);

  function SearchSelect() {
    var _this;

    _classCallCheck(this, SearchSelect);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(SearchSelect).call(this));

    _this.toggleClose = function (e) {
      if (!_this.state.deployed || _this.node.contains(e.target)) {
        return;
      }

      _this.toggleDeploy();
    };

    _this.toggleDeploy = function () {
      _this.setState({
        deployed: !_this.state.deployed
      });
    };

    _this.handleSearchChange = function (e) {
      _this.setState({
        searchField: e.target.value
      });
    };

    _this.filterOptions = function () {
      var filteredOptions = _this.getFilteredOptions();

      if (filteredOptions.length > 0) {
        return _this.renderOptions(filteredOptions);
      } else {
        return React.createElement("li", {
          className: "ter-search-select__options-list-item ter-search-select__options-list-item--inactive"
        }, "No results for ".concat(_this.state.searchField));
      }
    };

    _this.getFilteredOptions = function () {
      return _this.props.options.filter(function (option) {
        return option.toLowerCase().substring(0, _this.state.searchField.length) === _this.state.searchField;
      });
    };

    _this.renderOptions = function (options) {
      return options.map(function (option, index) {
        return React.createElement("li", {
          key: "".concat(option, "-").concat(index),
          onClick: function onClick() {
            return _this.handleSelect(option);
          },
          className: "ter-search-select__options-list-item"
        }, option);
      });
    };

    _this.toggleSelect = function () {
      _this.setState({
        deployed: !_this.state.deployed
      });
    };

    _this.handleSelect = function (option) {
      _this.setState({
        deployed: false
      });

      _this.props.handleSelect(option, _this.props.name);
    };

    _this.state = {
      deployed: false,
      searchField: ""
    };
    return _this;
  }

  _createClass(SearchSelect, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      document.addEventListener("mousedown", this.toggleClose, false);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      document.removeEventListener("mousedown", this.toggleClose, false);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return React.createElement("div", {
        ref: function ref(node) {
          return _this2.node = node;
        },
        className: "ter-search-select ".concat(this.state.deployed ? "is-open" : "")
      }, React.createElement("div", {
        className: "ter-search-select__selected",
        onClick: this.toggleDeploy
      }, this.props.selection || this.props.defaultText, React.createElement(Icon, {
        type: "open-caret-down-dark-16px",
        className: "ter-search-select__caret",
        size: "16px"
      })), React.createElement("ul", {
        className: "ter-search-select__options-list"
      }, React.createElement("input", {
        ref: function ref(input) {
          return input && input.focus();
        },
        type: "text",
        value: this.state.searchField,
        onChange: function onChange(e) {
          return _this2.handleSearchChange(e);
        },
        className: "ter-search-select__search-input",
        placeholder: "search"
      }), this.state.searchField ? this.filterOptions() : this.renderOptions(this.props.options)), React.createElement("div", null));
    }
  }]);

  return SearchSelect;
}(Component);

export default SearchSelect;