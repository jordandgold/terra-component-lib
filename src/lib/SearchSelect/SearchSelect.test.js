import React from "react";
import { shallow, mount } from "enzyme";
import SearchSelect from "./SearchSelect";

describe("SearchSelect", () => {
  let wrapper;
  let mockHandleSelect;

  beforeEach(() => {
    mockHandleSelect = jest.fn();
    const mockOptions = [
      "cats",
      "dogs",
      "turtles",
      "fish",
      "ferrets",
      "hamsters",
      "birds"
    ];

    wrapper = shallow(
      <SearchSelect
        defaultText="mock default text"
        selection={undefined}
        handleSelect={mockHandleSelect}
        options={mockOptions}
      />
    );
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("toggleDeploy", () => {
    it("should set state", () => {
      wrapper.instance().toggleDeploy();

      expect(wrapper.state().deployed).toEqual(true);
    });

    it("should call toggleDeploy on click", () => {
      const spy = jest.spyOn(wrapper.instance(), "toggleDeploy");
      wrapper.instance().forceUpdate();

      wrapper.find(".ter-search-select-label").simulate("click");

      expect(spy).toHaveBeenCalled();
    });

    it("should focus the input when opening", () => {
      const mockOptions = [
        "cats",
        "dogs",
        "turtles",
        "fish",
        "ferrets",
        "hamsters",
        "birds"
      ];

      wrapper = mount(
        <SearchSelect
          defaultText="mock default text"
          selection={undefined}
          handleSelect={mockHandleSelect}
          options={mockOptions}
        />
      );

      expect(
        document.activeElement[Object.keys(document.activeElement)[0]].type
      ).toEqual("input");
    });
  });

  describe("handleSearchChange", () => {
    it("should set state", () => {
      const mockEvent = {
        target: {
          value: "mock value"
        }
      };

      wrapper.instance().handleSearchChange(mockEvent);

      expect(wrapper.state().searchField).toEqual("mock value");
    });

    it("should call handleSearchChange on change", () => {
      const spy = jest.spyOn(wrapper.instance(), "handleSearchChange");
      wrapper.instance().forceUpdate();
      const mockEvent = { target: { value: "testing" } };

      wrapper.find("input").simulate("change", mockEvent);

      expect(spy).toHaveBeenCalledWith(mockEvent);
    });
  });

  describe("filterOptions", () => {
    it("should call getFilteredOptions", () => {
      const spy = jest.spyOn(wrapper.instance(), "getFilteredOptions");
      wrapper.instance().forceUpdate();

      wrapper.instance().filterOptions();

      expect(spy).toHaveBeenCalled();
    });

    it("should call renderOptions", () => {
      const spy = jest.spyOn(wrapper.instance(), "renderOptions");
      wrapper.instance().forceUpdate();

      wrapper.instance().filterOptions();

      expect(spy).toHaveBeenCalled();
    });

    it("should return", () => {
      const spy = jest.spyOn(wrapper.instance(), "filterOptions");
      wrapper.instance().forceUpdate();

      wrapper.instance().filterOptions();

      expect(spy).toReturn();
    });
  });

  describe("getFilteredOptions", () => {
    it("should return a filtered array", () => {
      const expected = ["fish", "ferrets"];
      wrapper.state().searchField = "f";

      const result = wrapper.instance().getFilteredOptions();

      expect(result).toEqual(expected);
    });
  });

  describe("renderOptions", () => {
    it("should return some JSX", () => {
      const mockOptions = ["cat", "dog", "fish"];
      const expected = [
        '<li class="select-options__items">cat</li>',
        '<li class="select-options__items">dog</li>',
        '<li class="select-options__items">fish</li>'
      ];

      const result = wrapper
        .instance()
        .renderOptions(mockOptions)
        .map(option => {
          return shallow(option).html();
        });

      expect(result).toEqual(expected);
    });
  });

  describe("toggleSelect", () => {
    it("should setState", () => {
      wrapper.instance().toggleSelect();

      expect(wrapper.state().deployed).toEqual(true);
    });
  });

  describe("handleSelect", () => {
    it("should set state", () => {
      wrapper.state().deployed = true;

      wrapper.instance().handleSelect("mock option");

      expect(wrapper.state().deployed).toEqual(false);
    });

    it("should call this.props.handleSelect", () => {
      wrapper.state().deployed = true;

      wrapper.instance().handleSelect("mock option");

      expect(mockHandleSelect).toHaveBeenCalledWith("mock option");
    });

    it("should call handleSelect on click", () => {
      wrapper = shallow(
        <SearchSelect
          defaultText="mock default text"
          selection={undefined}
          handleSelect={mockHandleSelect}
          options={["cats"]}
        />
      );

      const spy = jest.spyOn(wrapper.instance(), "handleSelect");

      wrapper.find(".select-options__items").simulate("click");

      expect(spy).toHaveBeenCalledWith("cats");
    });
  });
});
