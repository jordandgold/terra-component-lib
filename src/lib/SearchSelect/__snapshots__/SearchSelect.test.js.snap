// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSelect should match the snapshot 1`] = `
<div
  className="ter-search-select "
>
  <div
    className="ter-search-select__selected"
    onClick={[Function]}
  >
    mock default text
    <Icon
      className="ter-search-select__caret"
      size="16px"
      type="open-caret-down-dark-16px"
    />
  </div>
  <ul
    className="ter-search-select__options-list"
  >
    <input
      className="ter-search-select__search-input"
      onChange={[Function]}
      placeholder="search"
      type="text"
      value=""
    />
    <li
      className="ter-search-select__options-list-item"
      key="cats-0"
      onClick={[Function]}
    >
      cats
    </li>
    <li
      className="ter-search-select__options-list-item"
      key="dogs-1"
      onClick={[Function]}
    >
      dogs
    </li>
    <li
      className="ter-search-select__options-list-item"
      key="turtles-2"
      onClick={[Function]}
    >
      turtles
    </li>
    <li
      className="ter-search-select__options-list-item"
      key="fish-3"
      onClick={[Function]}
    >
      fish
    </li>
    <li
      className="ter-search-select__options-list-item"
      key="ferrets-4"
      onClick={[Function]}
    >
      ferrets
    </li>
    <li
      className="ter-search-select__options-list-item"
      key="hamsters-5"
      onClick={[Function]}
    >
      hamsters
    </li>
    <li
      className="ter-search-select__options-list-item"
      key="birds-6"
      onClick={[Function]}
    >
      birds
    </li>
  </ul>
  <div />
</div>
`;
